File: common/src/main/scala/edu/cdl/iot/common/security/EncryptionHelper.scala
package edu.cdl.iot.common.security

import java.security.SecureRandom
import java.util
import java.util.Base64

import javax.crypto.{Cipher, SecretKeyFactory}
import javax.crypto.spec.{IvParameterSpec, PBEKeySpec, SecretKeySpec}


class EncryptionHelper(val key: String,
                       val salt: String,
                       val mode: Int = Cipher.ENCRYPT_MODE,
                       val keyScheme: String = "SHA1") extends Serializable {
  private val factoryInstance = s"PBKDF2WithHmac${keyScheme}"
  private val cipherInstance = "AES/CBC/PKCS5PADDING"
  private val secretKeyType = "AES"
  private val ivCode = new Array[Byte](16)
  @transient lazy private val cipher = initCipher(key, salt, mode)

  def transform(input: Map[String, String]): Map[String, String] = input.map({
    case (x, d) =>
      transform(x) -> transform(d)
  })

  def transform(input: String): String = mode match {
    case Cipher.ENCRYPT_MODE => encrypt(input)
    case Cipher.DECRYPT_MODE => decrypt(input)
    case _ => throw new IllegalAccessException(
      s"Mode ${mode} not supported, only Encrypt (${Cipher.ENCRYPT_MODE}) and Decrypt (${Cipher.DECRYPT_MODE}) are supported"
    )
  }

  // Private "Ugly" methods
  private def encrypt(value: String): String = {
    val encrypted = cipher.doFinal(value.getBytes)
    val cipherWithIv = addIVToCipher(encrypted)
    Base64.getEncoder.encodeToString(cipherWithIv)
  }


  private def decrypt(encrypted: String): String = {
    val original = cipher.doFinal(Base64.getDecoder.decode(encrypted))
    val originalWithoutIv = util.Arrays.copyOfRange(original, 16, original.length)
    new String(originalWithoutIv)
  }


  private def initCipher(secretKey: String, salt: String, mode: Int) = {
    val factory = SecretKeyFactory.getInstance(factoryInstance)
    val spec = new PBEKeySpec(secretKey.toCharArray, salt.getBytes, 65536, 256)
    val tmp = factory.generateSecret(spec)
    val skeySpec = new SecretKeySpec(tmp.getEncoded, secretKeyType)
    val cipher = Cipher.getInstance(cipherInstance)
    // Generating random IV
    val random = new SecureRandom
    random.nextBytes(ivCode)
    cipher.init(mode, skeySpec, new IvParameterSpec(ivCode))
    cipher
  }

  private def addIVToCipher(encrypted: Array[Byte]) = {
    val cipherWithIv = new Array[Byte](ivCode.length + encrypted.length)
    System.arraycopy(ivCode, 0, cipherWithIv, 0, ivCode.length)
    System.arraycopy(encrypted, 0, cipherWithIv, ivCode.length, encrypted.length)
    cipherWithIv
  }
}

File: common/src/main/python/EncryptionHelper.py
import base64
import hashlib
from enum import Enum

from Crypto import Random
from Crypto.Cipher import AES


class EncryptionMode(Enum):
    ENCRYPT = 1
    DECRYPT = 2


class EncryptionHelper:
    # TODO: Use a proper salt with project GUID eventually
    def __init__(self, key, salt, mode=EncryptionMode.ENCRYPT, key_scheme='SHA1'):
        self.key = key
        self.salt = salt
        self.mode = mode
        self.key_scheme = key_scheme
        self.private_key = hashlib.pbkdf2_hmac(self.key_scheme, self.key.encode(), self.salt.encode(), 65536, 32)

        self.block_size = 16
        self.pad = lambda s: s + (self.block_size - len(s) % self.block_size) * chr(
            self.block_size - len(s) % self.block_size)
        self.unpad = lambda s: s[0:-ord(s[-1:])]
        self.iv = Random.new().read(AES.block_size)

    def get_private_key(self, secret_key, salt):
        return

    def transform(self, input):
        if self.mode == EncryptionMode.ENCRYPT:
            return self.__encrypt(input)
        elif self.mode == EncryptionMode.DECRYPT:
            return self.__decrypt(input)
        else:
            raise Exception(
                f'Mode {self.mode} not supported. only Encryption ({EncryptionMode.ENCRYPT}) and Decryption ({EncryptionMode.DECRYPT}) are supported')

    def __encrypt(self, message):
        private_key = self.private_key
        message = self.pad(message)
        cipher = AES.new(private_key, AES.MODE_CBC, self.iv)
        cipher_bytes = base64.b64encode(self.iv + cipher.encrypt(message))
        return bytes.decode(cipher_bytes)

    def __decrypt(self, encoded):
        private_key = self.private_key
        cipher_text = base64.b64decode(encoded)
        iv = cipher_text[:AES.block_size]
        cipher = AES.new(private_key, AES.MODE_CBC, iv)
        plain_bytes = self.unpad(cipher.decrypt(cipher_text[self.block_size:]));
        return bytes.decode(plain_bytes)
